lib/queue.js

```diff
@@ -830,16 +830,40 @@ class Queue extends Emitter {
       args[i] = arguments[i];
     }
 
-    return this._commandable().then((client) => {
-      // Get the sha for the script after we're ready to avoid a race condition
-      // with the lua script loader.
-      args[0] = lua.shas[name];
-
-      const promise = helpers.deferred();
-      args.push(promise.defer());
-      client.evalsha.apply(client, args);
-      return promise;
-    });
+    return this._commandable()
+      .then((client) => {
+        // Get the sha for the script after we're ready to avoid a race
+        // condition with the lua script loader.
+        args[0] = lua.shas[name];
+
+        const promise = helpers.deferred();
+        args.push(promise.defer());
+        client.evalsha.apply(client, args);
+        return promise;
+      })
+      .catch((err) => {
+        // In cases where the Redis server fails over, the script cache is
+        // cleared (even when the cache itself is still intact.) We detect this
+        // condition and reload the scripts.
+        if (
+          err.name === 'ReplyError' &&
+          err.message === 'NOSCRIPT No matching script. Please use EVAL.'
```

Comment:
this isn't well-documented, but `node_redis` errors have a `code` field:
```suggestion
          err.code === 'NOSCRIPT'
```

```
ReplyError: NOSCRIPT No matching script. Please use EVAL.
    at parseError (/private/tmp/redis/node_modules/redis-parser/lib/parser.js:193:12)
    at parseType (/private/tmp/redis/node_modules/redis-parser/lib/parser.js:303:14) {
  command: 'EVALSHA',
  args: [ 'asdkjhdfg', 0 ],
  code: 'NOSCRIPT'
}```

Reply:
this is excellent; I was nervous about including the entire string, because that could easily change in the future.